File: ocaml/checker.ml
open Datatypes
open Prelude
open Errors

let rec subst name mu tau =
  match tau with
  | Lit _ -> tau
  | Var name' -> if name = name' then mu else tau
  | Symtree xs -> Symtree (List.map (subst name mu) xs)
  | Hole -> tau

let multisubst substs tau =
  let salt = gensym () in
  Env.fold    (fun name _  -> subst name (Var (name ^ salt))) substs tau
  |> Env.fold (fun name mu -> subst (name ^ salt) mu) substs
  |> Env.fold (fun name _  -> subst (name ^ salt) (Var name)) substs

let prune substs tau =
  match tau with
  | Var name ->
    if Env.mem name substs then
      Env.find name substs
    else tau
  | _ -> tau

let rec matches substs patt tau =
  let err = MatchError (patt, tau) in
  let omega = prune substs patt in
  match omega, tau with
  | Var name, _ -> Env.add name tau substs
  | Lit a, Lit b -> if a = b then substs else raise err
  | Symtree xs, Symtree ys ->
    if List.length xs <> List.length ys then raise err
    else List.fold_left2 matches substs xs ys
  | Hole, _ -> substs
  | _, _ -> raise err

let itMatches substs patt tau =
  try ignore (matches substs patt tau); true
  with MatchError _ -> false

let getMatch substs patt tau =
  try Some (matches substs patt tau);
  with MatchError _ -> None

let rec occurs tau name =
  match tau with
  | Var name'  -> name = name'
  | Lit _      -> false
  | Symtree xs -> List.exists (fun x -> occurs x name) xs
  | _          -> false

let even fi tau =
  if fi <> tau then raise (UnificationError (fi, tau))
  else ()

let lookup (ctx : inferenceRule Env.t) name =
  match Env.find_opt name ctx with
  | Some v -> v
  | _      -> raise (NotDefinedError name)

(* Compatibility with OCaml 4.05
   From: https://github.com/ocaml/ocaml/blob/trunk/stdlib/list.ml *)
let rec findMap f = function
  | [] -> None
  | x :: l ->
    begin match f x with
      | Some _ as result -> result
      | None -> findMap f l
    end

let rec getBound (bound : term list) tau =
  let formula = findMap (fun x -> getMatch Env.empty x tau) bound in
  let vars =
    begin match formula with
      | Some substs ->
        List.map
          (function
          | _, Var name -> name
          | _, omega    -> raise (ExpectedVariable omega))
          (Env.bindings substs)
      | None -> []
    end in
  match tau with
  | Symtree xs -> vars @ List.concat (List.map (getBound bound) xs)
  | _          -> vars

let isVar : term -> bool = function
  | Var _ -> true
  | _     -> false

let checkSubst (bound : term list) (substs : env) tau =
  let vars = getBound bound tau in
  Env.iter (fun name omega ->
    if List.mem name vars && not (isVar omega) then
      raise (ReplacingBoundWithConstant (name, omega))
    else match omega with
    | Var name' ->
      if List.mem name' vars then
        raise (ReplacingWithBound (name, name'))
      else ()
    | _ -> ())
    substs

let sorry tree tau =
  match tree with
  | Sorry name -> Printf.printf "%s: expected “%s”" name (showTerm tau)
  | _          -> ()

let rec infer (ctx : inferenceRule Env.t) bound tree =
  let statement = lookup ctx tree.edge in
  let v1 = List.length tree.children in
  let v2 = List.length statement.premises in
  if v1 <> v2 then raise (ApplicationMismatch (tree.edge, v1, v2)) else ();
  List.iter2 (fun premise child ->
    checkSubst bound tree.substitutions premise;
    let expected = multisubst tree.substitutions premise in
    match child with
    | Sorry _ -> sorry child expected
    | Proof x -> infer ctx bound x |> even expected)
    statement.premises tree.children;
  checkSubst bound tree.substitutions statement.conclusion;
  multisubst tree.substitutions statement.conclusion

let check ctx bound tau : derivation -> unit = function
  | Sorry x -> sorry (Sorry x) tau
  | Proof x -> even (infer ctx bound x) tau

File: python/prover/checker.py
from prover.datatypes import *
from prover.prelude import *
from prover.errors import *

from functools import partial
from typing import List, Dict

def subst(name : Name, μ : Term, τ : Term) -> Term:
    if isinstance(τ, Lit):
        return τ
    elif isinstance(τ, Var):
        if τ.name == name:
            return μ
        else:
            return τ
    elif isinstance(τ, Symtree):
        return Symtree(maplist(partial(subst, name, μ), τ.children))
    elif isinstance(τ, Hole):
        return τ

def multisubst(substitutions : Dict[Name, Term], τ : Term) -> Term:
    salt = gensym()
    for name, _ in substitutions.items():
        τ = subst(name, Var(name + salt), τ)

    for name, μ in substitutions.items():
        τ = subst(name + salt, μ, τ)

    for name, _ in substitutions.items():
        τ = subst(name + salt, Var(name), τ)
    return τ

def prune(substs : Dict[Name, Term], τ : Term):
    if isinstance(τ, Var):
        if τ.name in substs:
            return substs[τ.name]
        else:
            return τ
    else:
        return τ

def match(substs : Dict[Name, Term], patt : Term, τ : Term) -> bool:
    σ = prune(substs, patt)
    if isinstance(σ, Var):
        substs[σ.name] = τ
        return True
    elif isinstance(σ, Lit) and isinstance(τ, Lit):
        return σ.name == τ.name
    elif isinstance(σ, Symtree) and isinstance(τ, Symtree):
        if len(σ.children) != len(τ.children):
            return False
        for δσ, δτ in zip(σ.children, τ.children):
            if not match(substs, δσ, δτ):
                return False
        return True
    elif isinstance(σ, Hole):
        return True
    else:
        return False

def occurs(τ : Term, name : Name):
    if isinstance(τ, Var):
        return τ.name == name
    elif isinstance(τ, Lit):
        return False
    elif isinstance(τ, Symtree):
        for σ in τ.children:
            if occurs(σ, name):
                return True
        return False
    else:
        return False

def even(φ : Term, ψ : Term):
    if φ != ψ:
        raise UnificationError(φ, ψ)

def lookup(ctx : Dict[Name, InferenceRule], name : Name) -> InferenceRule:
    if name in ctx:
        return ctx[name]
    else:
        raise NotDefinedError(name)

def getbound(bound : List[Term], τ : Term) -> List[Name]:
    vars = []

    for formula in bound:
        substs = {}
        if match(substs, formula, τ):
            for name, σ in substs.items():
                if not isinstance(σ, Var):
                    raise VerificationError("“%s” expected to be a variable" % σ)
            vars = maplist(lambda σ: σ.name, substs.values())
            break

    if isinstance(τ, Symtree):
        for child in τ.children:
            vars.extend(getbound(bound, child))
    return vars

def checksubst(bound : List[Term], substitutions : Dict[Name, Term], τ : Term):
    BV = getbound(bound, τ)
    for name, σ in substitutions.items():
        if name in BV and not isinstance(σ, Var):
            raise VerificationError(
                "cannot replace bound variable “{0}” with a constant “{1}”".format(
                    name, σ
                )
            )
        elif isinstance(σ, Var) and σ.name in BV:
            raise VerificationError(
                "cannot replace “{0}” with “{1}”, because “{1}” is bound".format(
                    name, σ.name
                )
            )

def sorry(arg : Sorry, τ : Term):
    print("%s: expected “%s”" % (arg.name, τ))

def infer(ctx : Dict[Name, InferenceRule], bound : List[Term], tree : Proof) -> Term:
    statement = lookup(ctx, tree.edge)
    if len(tree.arguments) != len(statement.premises):
        raise VerificationError(
           "%s expects %d arguments, but got %d" % (
               tree.edge,
               len(statement.premises),
               len(tree.arguments)
           )
        )

    for premise, arg in zip(statement.premises, tree.arguments):
        checksubst(bound, tree.substitutions, premise)
        expected = multisubst(tree.substitutions, premise)

        if isinstance(arg, Lemma):
            even(expected, infer(ctx, bound, Proof(arg.name, [], {})))
        else:
            sorry(arg, expected)

    checksubst(bound, tree.substitutions, statement.conclusion)
    return multisubst(tree.substitutions, statement.conclusion)

def check(ctx : Dict[Name, InferenceRule], bound : List[Term],
          τ : Term, tree : Proof):
    even(infer(ctx, bound, tree), τ)